## Sales Prediction
- In this notebook we will explore how we managed to predict the sales of the pharmacy stores for the next 6 weeks and also the methods we used

### Importing neccessary libraries and modules

In [5]:
import pandas as pd
import warnings
import os 
os.chdir('..')
from scripts.pipeline import *
warnings.filterwarnings('ignore')

### Loading the 3 datasets

In [2]:
store_df=pd.read_csv('C:/Users/abenet/Desktop/data/Week 4/Technical Content/data/store.csv')
train_df=pd.read_csv('C:/Users/abenet/Desktop/data/Week 4/Technical Content/data/train.csv')
test_df=pd.read_csv('C:/Users/abenet/Desktop/data/Week 4/Technical Content/data/test.csv')

In [9]:
store_df

Store StoreType Assortment  CompetitionDistance  \
0         1         c          a               1270.0   
1         2         a          a                570.0   
2         3         a          a              14130.0   
3         4         c          c                620.0   
4         5         a          a              29910.0   
...     ...       ...        ...                  ...   
1110   1111         a          a               1900.0   
1111   1112         c          c               1880.0   
1112   1113         a          c               9260.0   
1113   1114         a          c                870.0   
1114   1115         d          c               5350.0   

      CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                           9.0                    2008.0       0   
1                          11.0                    2007.0       1   
2                          12.0                    2006.0       1   
3                           9.0                    2009.0       0   
4                           4.0                    2015.0       0   
...                         ...                       ...     ...   
1110                        6.0                    2014.0       1   
1111                        4.0                    2006.0       0   
1112                        0.0                       0.0       0   
1113                        0.0                       0.0       0   
1114                        0.0                       0.0       1   

      Promo2SinceWeek  Promo2SinceYear     PromoInterval  
0                 0.0              0.0          No Promo  
1                13.0           2010.0   Jan,Apr,Jul,Oct  
2                14.0           2011.0   Jan,Apr,Jul,Oct  
3                 0.0              0.0          No Promo  
4                 0.0              0.0          No Promo  
...               ...              ...               ...  
1110             31.0           2013.0   Jan,Apr,Jul,Oct  
1111              0.0              0.0          No Promo  
1112              0.0              0.0          No Promo  
1113              0.0              0.0          No Promo  
1114             22.0           2012.0  Mar,Jun,Sept,Dec  

[1115 rows x 10 columns]

In [10]:
train_df

Store  DayOfWeek       Date  Sales  Customers  Open  Promo  \
0            1          5 2015-07-31   5263        555     1      1   
1            2          5 2015-07-31   6064        625     1      1   
2            3          5 2015-07-31   8314        821     1      1   
3            4          5 2015-07-31  13995       1498     1      1   
4            5          5 2015-07-31   4822        559     1      1   
...        ...        ...        ...    ...        ...   ...    ...   
1017204   1111          2 2013-01-01      0          0     0      0   
1017205   1112          2 2013-01-01      0          0     0      0   
1017206   1113          2 2013-01-01      0          0     0      0   
1017207   1114          2 2013-01-01      0          0     0      0   
1017208   1115          2 2013-01-01      0          0     0      0   

           StateHoliday  SchoolHoliday  Year  Month  Day  Weekdays  Weekends  \
0            No Holiday              1  2015      7   31      True     False   
1            No Holiday              1  2015      7   31      True     False   
2            No Holiday              1  2015      7   31      True     False   
3            No Holiday              1  2015      7   31      True     False   
4            No Holiday              1  2015      7   31      True     False   
...                 ...            ...   ...    ...  ...       ...       ...   
1017204  Public Holiday              1  2013      1    1      True     False   
1017205  Public Holiday              1  2013      1    1      True     False   
1017206  Public Holiday              1  2013      1    1      True     False   
1017207  Public Holiday              1  2013      1    1      True     False   
1017208  Public Holiday              1  2013      1    1      True     False   

               Holidays  
0           Not Holiday  
1           Not Holiday  
2           Not Holiday  
3           Not Holiday  
4           Not Holiday  
...                 ...  
1017204  New Year's Day  
1017205  New Year's Day  
1017206  New Year's Day  
1017207  New Year's Day  
1017208  New Year's Day  

[1017209 rows x 15 columns]

In [8]:
test_df

Id  Store  DayOfWeek       Date  Open  Promo StateHoliday  \
0          1      1          4 2015-09-17   1.0      1   No Holiday   
1          2      3          4 2015-09-17   1.0      1   No Holiday   
2          3      7          4 2015-09-17   1.0      1   No Holiday   
3          4      8          4 2015-09-17   1.0      1   No Holiday   
4          5      9          4 2015-09-17   1.0      1   No Holiday   
...      ...    ...        ...        ...   ...    ...          ...   
41083  41084   1111          6 2015-08-01   1.0      0   No Holiday   
41084  41085   1112          6 2015-08-01   1.0      0   No Holiday   
41085  41086   1113          6 2015-08-01   1.0      0   No Holiday   
41086  41087   1114          6 2015-08-01   1.0      0   No Holiday   
41087  41088   1115          6 2015-08-01   1.0      0   No Holiday   

       SchoolHoliday  Year  Month  Day  Weekdays  Weekends     Holidays  
0                  0  2015      9   17      True     False  Not Holiday  
1                  0  2015      9   17      True     False  Not Holiday  
2                  0  2015      9   17      True     False  Not Holiday  
3                  0  2015      9   17      True     False  Not Holiday  
4                  0  2015      9   17      True     False  Not Holiday  
...              ...   ...    ...  ...       ...       ...          ...  
41083              0  2015      8    1     False      True  Not Holiday  
41084              0  2015      8    1     False      True  Not Holiday  
41085              0  2015      8    1     False      True  Not Holiday  
41086              0  2015      8    1     False      True  Not Holiday  
41087              1  2015      8    1     False      True  Not Holiday  

[41088 rows x 14 columns]

---
- In the next cell, there might be a little confusion. I managed to make a class that can somewhat act like a pipeline and finish the series of steps that is needed for the pre-processing. This `DataUpdate` class does many things. The methods used inside this class are
    1. `date_generator`- This method inside the class generates neccessary column needed for the prediction. columns like the `Year`,`Month`, `Holidays`,etc.....
    2. `missing_handler`- This method is self-explanatory but this one fills the numerical columns that have null values with zero and the categorical columns with **No Promo**.(this is because the column `PromoInterval` is the only categorical column and it is appropriate to fill it with this value)
    3. `merge`- This method gets two datasets as a parameter and returns the merged version of those data.
    4. `encoder`- This method generally changes the categorical columns found on the train and test datasets and changes them to numbers using the **Label Encoder**. This make the model understand the categorical values.
    5. `scaler`- This method scales big numerical columns like `CompetitionDistance` to smaller number so it doesn't skew the prediction of the model. **Standard Scaler** is used for the scaling task.
    6. `transform`- This method basically runs all the methods mentioned above and returns the updated, cleaned and processed data. Running this method only is enough to execute all the 5 methods mentioned above.
- The class works by getting the specific store information from the `store.csv` file and merges it to the `train.csv` or `test.csv` file depending on the circumstance. Running this class for the train dataset and test dataset will make the columns in both data equal. Some columns that need to be removed are:
    1. `Sales`- This is the target column and the column we want to predict. So this column needs to be removed from the train dataset.
    2. `Customers`- This column is not found on the test dataset so we don't have to train the model using this column. Eventhough this column has the highest correlation with sales, it cannot be used for prediction.
    3. `Id`- This column is found on the test dataset and is only needed after we finished the sales prediction so it will be added in the final stage of the prediction.  
---

In [4]:
ana=DataUpdate(store_df,train_df)
dr=ana.transform()
dr

Store  StoreType  Assortment  CompetitionDistance  \
0            1          2           0            -0.537724   
1            1          2           0            -0.537724   
2            1          2           0            -0.537724   
3            1          2           0            -0.537724   
4            1          2           0            -0.537724   
...        ...        ...         ...                  ...   
1017204   1115          3           2            -0.008558   
1017205   1115          3           2            -0.008558   
1017206   1115          3           2            -0.008558   
1017207   1115          3           2            -0.008558   
1017208   1115          3           2            -0.008558   

         CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                         0.950877                  0.681904       0   
1                         0.950877                  0.681904       0   
2                         0.950877                  0.681904       0   
3                         0.950877                  0.681904       0   
4                         0.950877                  0.681904       0   
...                            ...                       ...     ...   
1017204                  -1.150188                 -1.464856       1   
1017205                  -1.150188                 -1.464856       1   
1017206                  -1.150188                 -1.464856       1   
1017207                  -1.150188                 -1.464856       1   
1017208                  -1.150188                 -1.464856       1   

         Promo2SinceWeek  Promo2SinceYear  PromoInterval  ...  Open Promo  \
0              -0.760097        -1.001128              3  ...     1     1   
1              -0.760097        -1.001128              3  ...     1     1   
2              -0.760097        -1.001128              3  ...     1     1   
3              -0.760097        -1.001128              3  ...     1     1   
4              -0.760097        -1.001128              3  ...     1     1   
...                  ...              ...            ...  ...   ...   ...   
1017204         0.675567         0.999118              2  ...     1     0   
1017205         0.675567         0.999118              2  ...     1     0   
1017206         0.675567         0.999118              2  ...     1     0   
1017207         0.675567         0.999118              2  ...     1     0   
1017208         0.675567         0.999118              2  ...     0     0   

         StateHoliday  SchoolHoliday  Year  Month  Day  Weekdays  Weekends  \
0                   2              1  2015      7   31      True     False   
1                   2              1  2015      7   30      True     False   
2                   2              1  2015      7   29      True     False   
3                   2              1  2015      7   28      True     False   
4                   2              1  2015      7   27      True     False   
...               ...            ...   ...    ...  ...       ...       ...   
1017204             2              1  2013      1    5     False      True   
1017205             2              1  2013      1    4      True     False   
1017206             2              1  2013      1    3      True     False   
1017207             2              1  2013      1    2      True     False   
1017208             3              1  2013      1    1      True     False   

         Holidays  
0               8  
1               8  
2               8  
3               8  
4               8  
...           ...  
1017204         8  
1017205         8  
1017206         8  
1017207         8  
1017208         7  

[1017209 rows x 24 columns]

### Model Training

---
* We can see that the above final dataset is a merged, scaled and encoded data. We can use this train the model. I used **Random Forest Regressor** to predict the sales.
---

In [6]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split,KFold,cross_val_score
X=dr.drop(['Sales','Customers','Date'],axis=1)
y=dr['Sales']
model=RandomForestRegressor(n_estimators=100, max_depth=15,min_samples_split=15, 
                               min_samples_leaf=10, n_jobs=-1, random_state=42)
X_train_split, X_val_split, y_train_split, y_val_split = train_test_split(X,y, test_size=0.3, random_state=42)
model.fit(X_train_split,y_train_split)

RandomForestRegressor(max_depth=15, min_samples_leaf=10, min_samples_split=15,
                      n_jobs=-1, random_state=42)

---
- The parameters I used can be seen above. I did the hyperparameter tuning by myself by using a **trail-and-error approach** because I wan unable to do any cross validation with this much data. The cross validation cells run for more than 25 minutes and even when they finished they bring up an error. So I adjusted the hyperparameters myself. The most influential parameter I examined regarding the model's performance was the `max_depth` parameter. I continously increased it and the r-squared score was also increasing. but I stopped at a value of 15 because at **max_depth=20**, the r-squared score for the model became **0.91** and when an r-2 score is this big, it is an indication of **overfitting**. the r-2 score and the mean absolute error I have found with the hyperparameters above can be shown below.
---

### Metrics 

In [7]:
from sklearn.metrics import r2_score,mean_absolute_error
y_pred=model.predict(X_val_split)
print(f"The r-2 score is {r2_score(y_val_split,y_pred)}")
print(f"The mean absolute error is {mean_absolute_error(y_val_split,y_pred)}")


The r-2 score is 0.8518049783088034
The mean absolute error is 944.8152688632964


---
* r-2 score of **0.85** is decent considering the size of the data. The mean aboslute error is **944.8**. the mae score indicates that on average the model's prediction is off by 944.8 units(in this case dollars). That is a huge difference and improvements should be made on the training of the data. When seeing that most of the stores have a sales in thousands, a mean absolute error of 944.8 units is fairly bad. But with the time I was given, this is the best I can provide.

* On the next cell, I run the `DataUpdate` class on the store and test dataset to transform the test dataset
---

In [6]:
da_2=DataUpdate(store_df,test_df)
test_cleaned=da_2.transform()
test_cleaned

Store  StoreType  Assortment  CompetitionDistance  \
0          1          2           0            -0.527160   
1          1          2           0            -0.527160   
2          1          2           0            -0.527160   
3          1          2           0            -0.527160   
4          1          2           0            -0.527160   
...      ...        ...         ...                  ...   
41083   1115          3           2             0.037848   
41084   1115          3           2             0.037848   
41085   1115          3           2             0.037848   
41086   1115          3           2             0.037848   
41087   1115          3           2             0.037848   

       CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                       1.084391                  0.766220       0   
1                       1.084391                  0.766220       0   
2                       1.084391                  0.766220       0   
3                       1.084391                  0.766220       0   
4                       1.084391                  0.766220       0   
...                          ...                       ...     ...   
41083                  -1.051128                 -1.303941       1   
41084                  -1.051128                 -1.303941       1   
41085                  -1.051128                 -1.303941       1   
41086                  -1.051128                 -1.303941       1   
41087                  -1.051128                 -1.303941       1   

       Promo2SinceWeek  Promo2SinceYear  PromoInterval  ...  Open  Promo  \
0            -0.876652        -1.176605              3  ...   1.0      1   
1            -0.876652        -1.176605              3  ...   1.0      1   
2            -0.876652        -1.176605              3  ...   1.0      1   
3            -0.876652        -1.176605              3  ...   1.0      1   
4            -0.876652        -1.176605              3  ...   0.0      0   
...                ...              ...            ...  ...   ...    ...   
41083         0.483242         0.850082              2  ...   1.0      1   
41084         0.483242         0.850082              2  ...   1.0      1   
41085         0.483242         0.850082              2  ...   1.0      1   
41086         0.483242         0.850082              2  ...   0.0      0   
41087         0.483242         0.850082              2  ...   1.0      0   

      StateHoliday  SchoolHoliday  Year  Month  Day  Weekdays  Weekends  \
0                0              0  2015      9   17      True     False   
1                0              0  2015      9   16      True     False   
2                0              0  2015      9   15      True     False   
3                0              0  2015      9   14      True     False   
4                0              0  2015      9   13     False      True   
...            ...            ...   ...    ...  ...       ...       ...   
41083            0              1  2015      8    5      True     False   
41084            0              1  2015      8    4      True     False   
41085            0              1  2015      8    3      True     False   
41086            0              1  2015      8    2     False      True   
41087            0              1  2015      8    1     False      True   

       Holidays  
0             1  
1             1  
2             1  
3             1  
4             1  
...         ...  
41083         1  
41084         1  
41085         1  
41086         1  
41087         1  

[41088 rows x 23 columns]

### Sales Prediction

In [2]:
test_id=pd.DataFrame(test_cleaned['Id'])
test_cleaned.drop(['Id','Date'],axis=1,inplace=True)
pred=pd.DataFrame(model.predict(test_cleaned))
test_id['Sales_Prediction']=pred[0]
test_id

Sales_Prediction     Id
0           5498.201776      1
1           5574.503961    857
2           5716.376821   1713
3           8533.671380   2569
4              0.000000   3425
...                 ...    ...
41083      22666.770214  37664
41084      22698.428738  38520
41085      24046.680693  39376
41086          0.000000  40232
41087       6389.665094  41088

[41088 rows x 2 columns]

* The `Id` is not in order because the .merge function changed the way the rows are ordered. But, every day of the predicted week and the store can be found by referring to the corresponding Id on the test dataset.

### Saving the model

In [18]:
import joblib
from datetime import datetime
timestamp = datetime.now().strftime('%d-%m-%Y-%H-%M-%S')
filename = f'model_{timestamp}.pkl'

joblib.dump(model, filename)
print(f'Model saved as {filename}')


Model saved as model_22-09-2024-16-40-36.pkl


### Saving the file that has the predicted sales

In [13]:
pred.to_csv('C:/Users/abenet/Desktop/data/sales_prediction.csv')